### PR TITLE
usb: dfu: refactor dfu class implementation

### PR DESCRIPTION
**Pull request by mrrosen**
From https://github.com/zephyrproject-rtos/zephyr/pull/30562, some smaller fixes for USB DFU class implementation were included in this PR that have been separated out in this PR as separate commits:

1. Implement appDETACH timer as required by USB DFU 1.1 specification (Section 5.1) so if the host fails to perform a USB reset within a certain windows, the device will return to appIDLE and future USB resets will not cause the device to incorrectly transition into DFU mode
2. Move the time when the device changes the descriptors to the DFU mode descriptors to when the host sends a USB reset rather than after receiving a DFU_DETACH command since while in appDETACH, the device is still in Run-time mode and should still return the Run-time descriptors if queried.
3. Refactor the code to allow for separate VID and PID for the DFU mode descriptor. At very least, separate PID is required for DFU mode as some host OS's will cache descriptors depending on the VID and PID combination. Using the same VID and PID for both the Run-time and DFU mode descriptors has been observed to cause issues with USB DFU on Windows 10 where the device is not found after USB reset due to descriptor caching.

### CODE REVIEW

**Comment by mrrosen** [comment]
@jfischer-no There was a bit of open discussion about the new VID/PID, I did a bit of what you had suggested as part of these commits but Im not sure we reached a final conclusion on whether there should be a separate VID, and what if anything cmake for DFU should check; since this is probably the most important fix in this commit due to it causing the most issues (again, Windows 10, dfu-utl 0.9 will always fail any DFU transaction the first time due to descriptor caching)

File: subsys/usb/class/dfu/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_sources(usb_dfu.c)
+
+if(CONFIG_USB_DEVICE_VID EQUAL CONFIG_USB_DEVICE_DFU_VID)
```
Comment (jfischer-no) [inline subsys/usb/class/dfu/CMakeLists.txt:5]: Please remove everything about USB_DEVICE_DFU_VID, it is not necessary to change VID.

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: subsys/usb/class/dfu/CMakeLists.txt (hunk 1 above)
Comment (mrrosen) [reply subsys/usb/class/dfu/CMakeLists.txt]: Done

**Comment by mrrosen** [comment]
@jfischer-no Any other changes need to be made to this PR?

**Review by jfischer-no — APPROVED** [review]

**Review by nvlsianpu — APPROVED** [review]

**Pull request merged**